instrumentation/compose/src/main/kotlin/io/opentelemetry/instrumentation/compose/ComposeClickEventGenerator.kt

```diff
@@ -0,0 +1,220 @@
+        queue.addFirst(owner.root)
+        var target: LayoutNode? = null
+
+        while (queue.isNotEmpty()) {
+            val node = queue.removeFirst()
+            if (node.isPlaced && hitTest(node, x, y)) {
+                target = node
+            }
+
+            queue.addAll(node.zSortedChildren.asMutableList())
+        }
+        return target
+    }
+
+    private fun isValidClickTarget(node: LayoutNode): Boolean {
+        var isClickable = false
+
+        for (info in node.getModifierInfo()) {
+            val modifier = info.modifier
+            if (modifier is SemanticsModifier) {
+                with(modifier.semanticsConfiguration) {
+                    isClickable = contains(SemanticsActions.OnClick)
+                }
+            } else {
+                val className = modifier::class.qualifiedName
+                isClickable = (
+                    className == CLASS_NAME_CLICKABLE_ELEMENT ||
+                        className == CLASS_NAME_COMBINED_CLICKABLE_ELEMENT ||
+                        className == CLASS_NAME_TOGGLEABLE_ELEMENT
+                )
+            }
+        }
+
+        return isClickable
+    }
+
+    private fun getNodeName(node: LayoutNode): String? {
+        var className: String? = null
+        for (info in node.getModifierInfo()) {
+            val modifier = info.modifier
+            if (modifier is SemanticsModifier) {
+                with(modifier.semanticsConfiguration) {
+                    val onClickSemanticsConfiguration = getOrNull(SemanticsActions.OnClick)
+                    if (onClickSemanticsConfiguration != null) {
+                        val accessibilityActionLabel = onClickSemanticsConfiguration.label
+                        if (accessibilityActionLabel != null) {
+                            return accessibilityActionLabel
+                        }
+                    }
+
+                    val contentDescriptionSemanticsConfiguration =
+                        getOrNull(SemanticsProperties.ContentDescription)
+                    if (contentDescriptionSemanticsConfiguration != null) {
+                        val contentDescription =
+                            contentDescriptionSemanticsConfiguration.getOrNull(0)
+                        if (contentDescription != null) {
+                            return contentDescription
+                        }
+                    }
+                }
+            } else {
+                className = modifier::class.qualifiedName
+            }
+        }
+
+        return className
+    }
+
+    private fun hitTest(
+        node: LayoutNode,
+        x: Float,
+        y: Float,
+    ): Boolean {
+        val bounded =
+            composeLayoutNodeUtil.getLayoutNodeBoundsInWindow(node)?.let { bounds ->
+                x >= bounds.left && x <= bounds.right && y >= bounds.top && y <= bounds.bottom
+            } == true
+
+        return bounded && isValidClickTarget(node)
+    }
```

Comment:
Wow, kinda surprised at how much work it was to get the target! Thanks for taking that on. I think this class has gotten a little long and has multiple responsibilities. Would it be possible to factor out the target finding capabilities into a separate collaborator class?